### PR TITLE
Adds Poseidon CRH w/ constraints

### PR DESCRIPTION
**Pull request by drewstone**
Following reference implementations from:

https://github.com/webb-tools/bulletproof-gadgets/blob/main/src/poseidon/mod.rs

### CODE REVIEW

**Comment by weikengchen** [comment]
You have a similar draft implementation currently experimental in `sponge`. 
https://github.com/arkworks-rs/sponge/blob/experimental/src/poseidon/mod.rs

It may be able if the work can be merged together. @Will-Lin4 is working on that part.

One thing:
- One observation from Fractal is that the cost of Poseidon is a little bit mysterious. For example, we may consider 3, 5, inverse:
```
PoseidonSboxGadget {
    Exponentiation3,
    Exponentiation5,
    Inverse,
}
```
It appears that sometimes, exp by 17 would be better for efficiency, for two reasons:
- 17 = 16 + 1, so not too hard to compute using fast powering.
- using exp by 17 allows us to spend fewer rounds.

And therefore, there is a slightly interesting cost model here.

**Comment by weikengchen** [comment]
Also, one of the big challenges currently for arkworks is the lack of Poseidon parameters for the common curves (no need for all). If you are interested in that part, it would be really appericated. 

We have an extended discussion on that in 
https://github.com/arkworks-rs/sponge/issues/8

**Comment by drewstone** [comment]
> It appears that sometimes, exp by 17 would be better for efficiency, for two reasons:
> 
> * 17 = 16 + 1, so not too hard to compute using fast powering.
> * using exp by 17 allows us to spend fewer rounds.
> 
> And therefore, there is a slightly interesting cost model here.

Sure thing, this can all be added to make it support different options. I can try to generalise it as well.

As far as merging goes, I'm definitely looking to have an implementation that is consistent with our existing one over at @webb-tools. This is a migration of sorts of the work we have there. I'm not sure if the sponge variant will yield the same hash but this is due to my lack of knowledge of truly what sponges are.

**Comment by drewstone** [comment]
> Also, one of the big challenges currently for arkworks is the lack of Poseidon parameters for the common curves (no need for all). If you are interested in that part, it would be really appreciated.
> 
> We have an extended discussion on that in
> [arkworks-rs/sponge#8](https://github.com/arkworks-rs/sponge/issues/8)

Are you thinking we have a repo of hardcoded parameters exportable for different curves or an implementation of the parameter generation for generic curves? I'd guess the latter is more robust and sustainable.

File: Cargo.toml [hunk 1]
```diff
@@ -35,7 +35,7 @@ derivative = { version = "2.0", features = ["use_core"] }
 tracing = { version = "0.1", default-features = false, features = [ "attributes" ], optional = true }
 
 [features]
-default = ["std"]
+default = ["std", "r1cs"]
```
Comment (drewstone) [inline Cargo.toml:38]:
```suggestion
default = ["std"]
```

**Comment by weikengchen** [comment]
I would request a review from Will since it is very likely going to merge with the Sponge somehow.

=> This suggests a need for the Sponge repo to provide interfaces for CRH in crypto-primitives, as it is needed for many downstream applications.

**Comment by drewstone** [comment]
@weikengchen I would suggest that if the two versions yield different hashes that we still have separate implementations. Otherwise, this will cause me to need to implement sponge variants as well in other applications (against bulletproofs, circom, etc.)

Having both is definitely useful also from visibility and example perspective.

**Review by weikengchen — APPROVED** [review]
The code looks good.

**Pull request merged**